src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptBlockStructureService.cs

```diff
@@ -0,0 +1,25 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Structure;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal abstract class VSTypeScriptBlockStructureService : BlockStructureService
```

Comment:
❗ A type in the `.Api` namespace cannot derive from an internal type

Reply:
This is the only way we can do this without coordinated TypeScript insertion. We already have a few cases like this.

Reply:
TypeScript already exports their BlockStructureService, so we can't export another one.

Reply:
Filed https://github.com/dotnet/roslyn/issues/58255 to follow up.

Reply:
It doesn't make any sense to declare an external API that cannot be referenced by the target project. It will be impossible for TypeScript to use this class.

Reply:
It actually works just fine.

Reply:
> This is the only way we can do this without coordinated TypeScript insertion

I don't understand this. Why can we not use the existing pattern of defining an interface that goes in the API layer? Prior to the TS insertion, this implementation of block structure service will not be active (no error), and after the TS insertion that provides an implementation of the interface the IDE will automatically switch from their current BlockStructureService to this one. Implementing this with correct External Access isolation is trivial and does not require a coordinated insertion. However, if we implement this in the current form and try to change it later, we will force a coordinated insertion requirement where none was otherwise present.

> It actually works just fine.

This is an error (oversight) in the analyzer, and cannot be treated as by design.

Reply:
TypeScript currently exports BlockStructureService:
```C#
[ExportLanguageService(typeof(BlockStructureService), StringConstants.TypeScriptLanguageName), Shared]
    internal sealed class TypeScriptBlockStructureService : VSTypeScriptBlockStructureService
```

If we export the service in ExternalAccess there will be two.

Reply:
> This is an error (oversight) in the analyzer, and cannot be treated as by design.

I'm not saying it's ideal, but it allows us make incremental improvements (reducing internal API exposure) without blocking on dual insertions.
There is a lot of `#pragma warning disable RS0035` in TypeScript.

Reply:
> If we export the service in ExternalAccess there will be two.

The one exported in ExternalAccess will only be exported if its own imports are satisfied. Since one of those imports will be an interface which is not implemented, it will be rejected from the MEF catalog. TypeScript's only insertion requirement will be to implement that interface in the same insertion that they remove the other export, which means there will never be a situation where more than one export is present.

Reply:
> There is a lot of `#pragma warning disable RS0035` in TypeScript.

This is why we should have just made separate assemblies. The analyzer isn't providing us with concrete protection.

Reply:
Separate assemblies wouldn't help. That would just mean Type Script has full IVT to everything.

Reply:
Will use the trick with invalid export.

Reply:
Done.